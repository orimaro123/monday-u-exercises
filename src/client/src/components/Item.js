import React, { useEffect, useState } from "react";
import ListControls from "./ListControls";

import { fetchAllItems, updateDoneTimestamp } from "../itemClient";
import editIcon from "../images/edit-icon.svg";
import saveIcon from "../images/save-icon.svg";
import deleteIcon from "../images/delete-icon.svg";

const ListItem = ({ item, itemToDelete, itemToEdit }) => {
  const [newName, setNewName] = useState(item.itemName);
  const [editSaveButtonIcon, setEditSaveButtonText] = useState(editIcon);
  const [itemWasEdited, setItemWasEdited] = useState(false);
  const [statusCompleteTime, setStatusCompleteTime] = useState("");
  const [hideClass, setHideClass] = useState("");
  const [readOnly, setReadOnly] = useState(true);
  const [decorateClass, setDecorateClass] = useState("");

  const newStatusHandler = async (e) => {
    if (e.target.checked) {
      setDecorateClass("decorate");

      const timestampNow = new Date();
      const timestampNowHours = timestampNow.getHours();
      timestampNow.setHours(timestampNowHours + 3);

      const timestampNowToDb = timestampNow
        .toISOString()
        .slice(0, 19)
        .replace("T", " ");
      const timestampNowToReact = timestampNowToDb.slice(10, 19);
      await updateDoneTimestamp(item.itemId, timestampNowToDb);
      setStatusCompleteTime(timestampNowToReact);
      setHideClass("");
      return;
    }
    setDecorateClass("");
    setHideClass("hide");
  };

  const editName = async () => {
    if (editSaveButtonIcon == editIcon) {
      setReadOnly(false);
      setEditSaveButtonText(saveIcon);
      return;
    }
    setReadOnly(true);
    setEditSaveButtonText(editIcon);
    let itemNewName = await itemToEdit(item.itemId, newName);

    setNewName(itemNewName);
  };

  return (
    <li className="list-item flex">
      {" "}
      <div className="check-box">
        <input type="checkBox" onChange={newStatusHandler} />
      </div>
      <input
        className={`list-item-text ${decorateClass}`}
        onChange={(e) => setNewName(e.target.value)}
        type="text"
        readOnly={readOnly}
        value={newName}
      />
      <div className={`status-complete-time  ${decorateClass} ${hideClass}`}>
        {statusCompleteTime}
      </div>
      {item.isPokemon ? (
        <div className={`pokemon-photo-div ${decorateClass}`}>
          {" "}
          <img src={item.pokemonData}></img>
        </div>
      ) : (
        <div className="space-div"></div>
      )}
      <div className="list-item-trash-div">
        <img
          className="list-item-trash-button"
          src={deleteIcon}
          onClick={() => itemToDelete(item.itemId ,item.itemName)}
        />
      </div>
      <div className="list-item-edit-div">
        <img
          className="list-item-edit-button"
          src={editSaveButtonIcon}
          onClick={editName}
        />
      </div>
    </li>
  );
};
export default ListItem;
