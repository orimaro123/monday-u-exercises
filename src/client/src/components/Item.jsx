import React, { useState, useEffect } from "react";

import { updateDoneTimestamp, updateStatusInDb } from "../services/itemClient";

import editIcon from "../images/edit-icon.svg";
import saveIcon from "../images/save-icon.svg";
import deleteIcon from "../images/delete-icon.svg";

const Item = ({ item, items, deleteItemAction, editItemAction, updateCheckBoxAction, checkBoxCheckRedux }) => {
 
 
// console.log(items)
  const [newName, setNewName] = useState(item.itemName);
  const [editSaveButtonIcon, setEditSaveButtonText] = useState(editIcon);

  const [statusCompleteTime, setStatusCompleteTime] = useState("");
  const [hideClass, setHideClass] = useState("");
  const [readOnly, setReadOnly] = useState(true);
  const [decorateClass, setDecorateClass] = useState("");
  const [checkBoxCheck, setCheckBoxCheck] = useState(item.status);

  const newStatusHandler = async (e) => {
    
    if (e.target.checked) {
      setDecorateClass("decorate");

      updateCheckBoxAction( item.itemId,e.target.checked)
      
      setCheckBoxCheck(true);
      const timestampNow = new Date();
      const timestampNowHours = timestampNow.getHours();
      timestampNow.setHours(timestampNowHours + 3);

      const timestampNowToDb = timestampNow
        .toISOString()
        .slice(0, 19)
        .replace("T", " ");
      const timestampNowToReact = timestampNowToDb.slice(10, 19);
     // await updateDoneTimestamp(item.itemId, timestampNowToDb);

     // await updateStatusInDb(item.itemId, e.target.checked);

      setStatusCompleteTime(`Done at ${timestampNowToReact}`);
      setHideClass("");
      return;
    } else {
     // await updateStatusInDb(item.itemId, false);
      updateCheckBoxAction(item.itemId,false)
      setCheckBoxCheck(false);
      setDecorateClass("");
      setHideClass("hide");
    }
  };

  const editName = async () => {
    if (editSaveButtonIcon === editIcon) {
      setReadOnly(false);
      setEditSaveButtonText(saveIcon);
      return;
    }
    setReadOnly(true);
    setEditSaveButtonText(editIcon);

    editItemAction(item.itemId, newName);
  };
  const deleteItemHandler = () => {
    deleteItemAction(item.itemId, item.itemName);
  };

  useEffect(() => {
    
  
    if (item.status) {
     
      setDecorateClass("decorate");

       const timeFromDb = item.doneAt
        .slice(0, 19)
        .replace("T", " ")
        .slice(10, 19);

      const timeHours = timeFromDb.slice(0, 3);
      const correctHours = Number(timeHours) + 3;

      const correctTimeToReact = timeFromDb.replace(
        timeHours,
        correctHours.toString() 
      );
      setStatusCompleteTime(`Done at ${correctTimeToReact}`);
    }
  }, [item]);

  return (
    <li className="list-item flex">
      <div className="check-box">
        <input
          type="checkBox"
          checked={item.status}
          onChange={newStatusHandler}
          onClick={() => updateCheckBoxAction( item.itemId,!item.status)} //setCheckBoxCheck(!checkBoxCheck)
        />
      </div>
      <input
        className={`list-item-text ${decorateClass}`}
        onChange={(e) => setNewName(e.target.value)}
        type="text"
        readOnly={readOnly}
        value={newName}
      />
      {item.isPokemon ? (
        <div className={`pokemon-photo-div ${decorateClass}`}>
          {" "}
          <img className="pokemon-photo" src={item.pokemonData} alt=""></img>
        </div>
      ) : (
        <div className="space-div"></div>
      )}
      <div className={`status-complete-time   ${hideClass}`}>
        {statusCompleteTime}
      </div>
      <div className="list-item-trash-div">
        <img
          className="list-item-trash-button"
          src={deleteIcon}
          alt="trash"
          onClick={deleteItemHandler}
        />
      </div>
      <div className="list-item-edit-div">
        <img
          className="list-item-edit-button"
          src={editSaveButtonIcon}
          alt={"edit icon"}
          onClick={editName}
        />
      </div>
    </li>
  );
};

export default Item;
