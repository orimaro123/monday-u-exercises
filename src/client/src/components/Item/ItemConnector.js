import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import Item from "./Item";

import {
  deleteItemAction,
  editItemAction,
  updateCheckBoxAction,
} from "../../redux/actions/itemsEntitiesActions";
import {
 
  getCheckBoxCheck,
} from "../../redux/selectors/itemsEntitiesSelectors";

const mapStateToProps = (state, ownProps) => {
  const checkBoxCheckRedux = getCheckBoxCheck(state);

  return { checkBoxCheckRedux };
};

const mapDispatchToProps = (dispatch, ownProps) => {
  return bindActionCreators(
    {
      deleteItemAction,
      editItemAction,
      updateCheckBoxAction,
    },
    dispatch
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(Item);
