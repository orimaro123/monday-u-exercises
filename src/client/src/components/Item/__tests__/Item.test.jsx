import { render, screen, cleanup } from "@testing-library/react";
import Item from "../Item";
import ShallowRenderer from "react-test-renderer/shallow";

it("should render item component", () => {
    const renderer = new ShallowRenderer()
  const item = {
    doneAt: null,
    id: 1,
    isPokemon: true,
    itemId: "2435268d-a0d0-49cb-bd59-6c2cacef552a",
    itemName: "Catch bulbasaur",
    pokemonData:
      "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
    pokemonId: 1,
    status: false,
  };
 
  const itemElement = renderer. render(<Item item={item}/>);
  expect(itemElement).toMatchSnapshot();
});

it("renders correctly", () => {
    const renderer = new ShallowRenderer();
 
    const item = {
       id: 3,
       text: "Pokemon",
    };
 
    const tree = renderer.render(<Item item={item} />);
    expect(tree).toMatchSnapshot();
 });
