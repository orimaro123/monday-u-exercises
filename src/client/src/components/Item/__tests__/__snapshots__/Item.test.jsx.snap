// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<li
  className="list-item flex"
>
  <div
    className="check-box"
  >
    <input
      onChange={[Function]}
      onClick={[Function]}
      type="checkBox"
    />
  </div>
  <input
    className="list-item-text "
    data-testid="item-3"
    onChange={[Function]}
    readOnly={true}
    type="text"
  />
  <div
    className="space-div"
  />
  <div
    className="status-complete-time   "
  >
    
  </div>
  <div
    className="list-item-trash-div"
  >
    <img
      alt="trash"
      className="list-item-trash-button"
      onClick={[Function]}
      src="delete-icon.svg"
    />
  </div>
  <div
    className="list-item-edit-div"
  >
    <img
      alt="edit icon"
      className="list-item-edit-button"
      onClick={[Function]}
      src="edit-icon.svg"
    />
  </div>
</li>
`;

exports[`should render item component 1`] = `
<li
  className="list-item flex"
>
  <div
    className="check-box"
  >
    <input
      checked={false}
      onChange={[Function]}
      onClick={[Function]}
      type="checkBox"
    />
  </div>
  <input
    className="list-item-text "
    data-testid="item-1"
    onChange={[Function]}
    readOnly={true}
    type="text"
    value="Catch bulbasaur"
  />
  <div
    className="pokemon-photo-div "
  >
     
    <img
      alt=""
      className="pokemon-photo"
      src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
    />
  </div>
  <div
    className="status-complete-time   "
  >
    
  </div>
  <div
    className="list-item-trash-div"
  >
    <img
      alt="trash"
      className="list-item-trash-button"
      onClick={[Function]}
      src="delete-icon.svg"
    />
  </div>
  <div
    className="list-item-edit-div"
  >
    <img
      alt="edit icon"
      className="list-item-edit-button"
      onClick={[Function]}
      src="edit-icon.svg"
    />
  </div>
</li>
`;
